Add support for FI_INT128/FI_UINT128 atomics.

jlbyrne-hpe opened the pull request:
Preserve existing types definitions and value for
FI_DATATYPE_LAST to preserve binary compatability.
Tested with shm provider.

top-level:
    configure.ac: Added check for 128-bit integer
    and atomics support.
    Makefile.am: Moved fi_atomic.3 to real_man_pages.
    man/fi_atomic.3.md, man/man3/fi_atomic.3: Documented
    new types and compiler support requirement.
    include/rdma/fi_domain.h: Added FI_INT128,
    FI_UINT128, and FI_DATATYPE_LAST_V2 after existing
    types.
    include/ofi_atomic.h: Replace FI_DATATYPE_LAST with
    FI_DATATYPE_LAST_V2 in arrays.
    src/fi_tostr.c: Added strings for new types.

prov/util: Added support for new types in sw atomics.
    util_atomic.c: Added support FI_INT128 and FI_UINT128
    to ofi_datatype_size(). Added support for 128-bit
    integer atomic library calls.

fabtests: Added support to test 128-bit atomics
    configure:ac: Added check for 128-bit integer support.
    common/shared.h: Added types/typedefs to support
    FI_INT128/FI_UINT128.
    common/shared.c: Added new types to size table.
    functional/rdm_atomic.c: Added parsing for int128
    and uint128 types.
    include/shared.h: Added types/typedefs to support
    new types.
    man/fabtests.7.md, man/man7/fabtests.: Documented
    new types and compiler support requirement.
    test_configs/shm/int128.test: Separate test config
    for new types.
    ubertest/config.c: Added parsing for new types.
    ubertest/ofi_atomic.c, ubertest/verify.c: Modified,
    added, macros for new types.
    ubertest/fabtest.h, ubertest/ofi_atomic.h: Replace
    FI_DATATYPE_LAST with FI_DATATYPE_LAST_V2 in arrays.

Signed-off-by: John Byrne <john.l.byrne@hpe.com>

shefty commented:
When a commit message has a list of items, that's usually an indication that the commit should be broken up into separate patches.

jlbyrne-hpe commented:
Okay.  Thoughts on how to break it up:

1.) Basic definition change.
2.) util_atomics and build changes (confiigure.ac).
3.) fi_atomic documentation changes (+Makefile.am change)
4.) rdm_atomic.c change + configure.ac
5.) ubertest changes
6.) fabtest documentation change.

I don't see much point in trying to finer than that. Tomorrow.

shefty commented:
Re-opening PR, so review comments don't get lost.  Just force push after updates.

jlbyrne-hpe commented:
We need to work out how to deal with presence/absence of compiler support for 128-bit integer math and atomics.  Points to keep in mind:

1.) Regardless of any compiler support, for providers that support 128-bit integer ops in HW we only need to make sure that the FI_INT128/FI_UINT128 defines exist and that ofi_datatype_size() returns the correct size for the new types, regardless of the compiler features.
2.) The fabtest rdm_atomic does not try to verify correctness of the operations, but that they simply complete without error and is largely independent of the compiler features. It does need datatype_to_size() routine to return the correct size in user space. 
3.) The fabtest ubertest tries to verify correctness of the operations atomic ops are performing. To do so it requires the compiler to support the arithmetic/comparison ops corresponding to the atomic ops performed by the provider. (e.g. It needs to locally perform a 128-bit integer add to compare the result with a FI_SUM 128-bit done by the provider.) 
4.) util_atomics.c gives providers that need to do atomics in SW a set of calls to do so. It relies on the the __atomic_xxx() extensions and the libatomic.so library. The gcc and libatomic I used supported same atomic ops for 128-bit integers as for 32 and 64 bit integers, so I modified the macros to support all the same ops and then I made sure ubertest could test them.

From the user perspective, using atomics in libfabrics has been about what the provider supports, but we now potentially have to worry about the test supports. If ubertest was built without 128-bit integer support, but the provider was,  it can test that a FI_INT128/FI_UINT128 completed without and error, but not verify the result. Our choices are to not run the test if we don't have the code path to verify it; run the test and inform the user of the limitation; or run the test differently depending on how the code was compiled.

As I said in (4), I implemented all the SW ops for FI_INT128/FI_UINT128 because I could, but from the HW perspective, there is only 1 op that will probably be cared about: FI_CSWAP. If we just implemented that, then the problem becomes much simpler. For SW support in the provider, instead of configure.ac testing for __int128 and the ability to compare-and-swap them,  we can define a struct of uint64_ts and write the configure.ac test for libfabric as:
```
    AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <stdint.h>]],                     
        [[struct { uint64_t i64[2]; } d, s, c;                                   
          __atomic_compare_exchange(&d, &c, &s, 0, __ATOMIC_SEQ_CST,             
                                    __ATOMIC_SEQ_CST);                           
         return 0;                                                               
        ]])],  
        [
            AC_MSG_RESULT(yes)
            AC_DEFINE(HAVE_BUILTIN_MM_128_CAS, 1, [Set to 1 to use built-in intrinsics memory model aware 128-bit compare-and-swap)
        ],
        [
            AC_MSG_RESULT(no)
        ]),                                                                  
```
The fabtests can be written to use the same struct and ubertest could use memcmp for equality in this case and wouldn't care about 128-bit arithmetic at all.

Thoughts?

I'm going to open a PR for the separate Makefile.am change, but otherwise I'm going to hold off on the other changes until we agree on the direction. The Travis tests produced a segv, so there might be some pushes to the branch to test that, please ignore them. The Appveyor tests died due to the use of __int128, so I'm going to wait on that, as well.

jlbyrne-hpe commented on Makefile.am:
```diff
@@ -224,6 +224,7 @@ real_man_pages = \
         man/man1/fi_info.1 \
         man/man1/fi_pingpong.1 \
         man/man1/fi_strerror.1 \
+        man/man3/fi_atomic.3 \
```
Submitted as : #6805

shefty commented:
> 1.) Regardless of any compiler support, for providers that support 128-bit integer ops
> in HW we only need to make sure that the FI_INT128/FI_UINT128 defines exist and that
> ofi_datatype_size() returns the correct size for the new types, regardless of the
> compiler features.

Agreed - Unlike the floating point definitions, the size here should be fixed at 128-bits.  We can actually hard code this size.

> 2.) The fabtest rdm_atomic does not try to verify correctness of the operations, but
> that they simply complete without error and is largely independent of the compiler
> features. It does need datatype_to_size() routine to return the correct size in user
> space.
> 3.) The fabtest ubertest tries to verify correctness of the operations atomic ops are
> performing. To do so it requires the compiler to support the arithmetic/comparison ops
> corresponding to the atomic ops performed by the provider. (e.g. It needs to locally
> perform a 128-bit integer add to compare the result with a FI_SUM 128-bit done by the
> provider.)

Fabtests can verify the results of the operation without having atomic support by the compiler.  It just needs an implementation of the operation.  Fabtests should be able to do this even if int128 itself is not supported, by emulating int128.  This is a valid use case for typedefs.

I would say that there is benefit for fabtest to verify int128 operations that a provider might perform even if there is no support for int128 directly.  I would also claim that full emulation might be a pain, so limiting testing to the most likely use cases would be acceptable.

> 4.) util_atomics.c gives providers that need to do atomics in SW a set of calls to do
> so. It relies on the the __atomic_xxx() extensions and the libatomic.so library. The
> gcc and libatomic I used supported same atomic ops for 128-bit integers as for 32 and
> 64 bit integers, so I modified the macros to support all the same ops and then I made
> sure ubertest could test them.

I agree that we want SW support for this.  And if we believe there's value in fabtests emulating int128, it makes sense for libfabric to emulate it as well (using the same basic implementation).

> From the user perspective, using atomics in libfabrics has been about what the provider
> supports, but we now potentially have to worry about the test supports. If ubertest was
> built without 128-bit integer support, but the provider was, it can test that a
> FI_INT128/FI_UINT128 completed without and error, but not verify the result. Our
> choices are to not run the test if we don't have the code path to verify it; run the
> test and inform the user of the limitation; or run the test differently depending on
> how the code was compiled.
> 
> As I said in (4), I implemented all the SW ops for FI_INT128/FI_UINT128 because I
> could, but from the HW perspective, there is only 1 op that will probably be cared
> about: FI_CSWAP. If we just implemented that, then the problem becomes much simpler.
> For SW support in the provider, instead of configure.ac testing for __int128 and the
> ability to compare-and-swap them, we can define a struct of uint64_ts and write the
> configure.ac test for libfabric as:
> 
>     AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <stdint.h>]],
>         [[struct { uint64_t i64[2]; } d, s, c;
>           __atomic_compare_exchange(&d, &c, &s, 0, __ATOMIC_SEQ_CST,
>                                     __ATOMIC_SEQ_CST);
>          return 0;
>         ]])],
>         [
>             AC_MSG_RESULT(yes)
>             AC_DEFINE(HAVE_BUILTIN_MM_128_CAS, 1, [Set to 1 to use built-in intrinsics
> memory model aware 128-bit compare-and-swap)
>         ],
>         [
>             AC_MSG_RESULT(no)
>         ]),
> 
> 
> The fabtests can be written to use the same struct and ubertest could use memcmp for
> equality in this case and wouldn't care about 128-bit arithmetic at all.
> 
> Thoughts?

I believe we want fabtests to emulate this, so that it can verify provider behavior.  And if fabtests already has to emulate the atomics, we copy the same logic into libfabric.  My preference is to emulate as many atomic ops as reasonable without int128.  With int128, I think we can emulate all ops.  If we have atomic-128 support, that should just replace specific emulated ops.

A staged implementation approach here is fine.  I think we need to start with a typedef to handle if int128 is defined or not, followed by adding support for the ops which are most useful.  CAS seems easy.  Others may not be.

j-xiong commented:
@jlbyrne-hpe This is supposed to be replaced by #6928 and #6929, is that right?

jlbyrne-hpe commented:
@j-xiong #6813 as well. Past time to close it.
